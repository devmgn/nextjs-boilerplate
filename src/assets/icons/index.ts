import Help from './help.svg?inline';
import Info from './info.svg?inline';

export { Help, Info };
