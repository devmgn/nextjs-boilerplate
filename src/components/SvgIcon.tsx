'use client';

import styled from 'styled-components';
import createShouldForwardProp from '@/utils/createShouldForwardProp';
import type { AnyComponent } from 'styled-components/dist/types';

const StyledSvgIcon = styled.svg.withConfig({
  shouldForwardProp: createShouldForwardProp('fontSize', 'color', 'rotate'),
})`
  display: inline-block;
  flex-shrink: 0;
  width: 1em;
  height: 1em;
  font-size: ${({ fontSize }) => (fontSize ? `${fontSize}px` : '24px')};
  color: ${({ color }) => color ?? 'inherit'};
  user-select: none;
  fill: currentColor;
  rotate: ${({ rotate }) => (rotate ? `${rotate}deg` : undefined)};
`;

type SvgIconProps = {
  as: AnyComponent;
} & Omit<React.ComponentPropsWithoutRef<typeof StyledSvgIcon>, 'as'>;

const SvgIcon: React.FC<SvgIconProps> = (props) => {
  return <StyledSvgIcon className="SvgIcon" focusable="false" aria-hidden="true" {...props} />;
};

export default SvgIcon;
