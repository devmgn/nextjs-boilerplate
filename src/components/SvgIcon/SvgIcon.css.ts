import { recipe } from '@vanilla-extract/recipes';
import { colorVariant } from '@/config';
import { numberToCssUnit } from '@/utils';
import type { RecipeVariants } from '@vanilla-extract/recipes';

export const svgIcon = recipe({
  base: {
    width: '1em',
    height: '1em',
  },
  variants: {
    size: {
      xs: { fontSize: numberToCssUnit(12) },
      sm: { fontSize: numberToCssUnit(16) },
      md: { fontSize: numberToCssUnit(20) },
      lg: { fontSize: numberToCssUnit(24) },
      xl: { fontSize: numberToCssUnit(32) },
    },
    color: colorVariant,
  },
  defaultVariants: {
    size: 'md',
    color: 'text',
  },
});

export type SvgIconVariants = RecipeVariants<typeof svgIcon>;
