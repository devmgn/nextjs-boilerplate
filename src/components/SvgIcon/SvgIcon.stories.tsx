import { FaceIcon } from "@radix-ui/react-icons";
import { expect, fn, userEvent, within } from "storybook/test";
import { SvgIcon } from ".";
import type { Meta, StoryObj } from "@storybook/nextjs-vite";

const meta = {
  component: SvgIcon,
  args: {
    icon: FaceIcon,
    label: "Face icon",
    size: "md",
    onClick: fn(),
  },
  argTypes: {
    size: {
      options: ["xs", "sm", "md", "lg", "xl"],
      control: { type: "inline-radio" },
    },
  },
} satisfies Meta<typeof SvgIcon>;

export default meta;
type Story = StoryObj<typeof SvgIcon>;

export const Default: Story = {};

export const Test: Story = {
  play: async ({ args, canvasElement }) => {
    const canvas = within(canvasElement);
    const icon = await canvas.findByLabelText("Face icon");
    await userEvent.click(icon);
    await expect(args.onClick).toHaveBeenCalled();
    await expect(icon).toHaveAttribute("aria-hidden", "true");
    await expect(icon).toHaveAttribute("focusable", "false");
  },
};
