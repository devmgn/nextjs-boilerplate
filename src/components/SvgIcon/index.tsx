import { cva } from 'class-variance-authority';
import { cn } from '@/lib/utils';
import type { VariantProps } from 'class-variance-authority';
import type { Merge } from 'type-fest';

const svgIconVariants = cva('size-[1em] fill-current text-current', {
  variants: {
    size: {
      xs: 'text-[12px]',
      sm: 'text-[16px]',
      md: 'text-[24px]',
      lg: 'text-[32px]',
      xl: 'text-[40px]',
    },
  },
  defaultVariants: {
    size: 'md',
  },
});

type SvgIconProps = Merge<
  { component: React.ElementType },
  React.HTMLAttributes<SVGSVGElement> & VariantProps<typeof svgIconVariants>
>;

export const SvgIcon: React.FC<SvgIconProps> = ({
  component: Component,
  className,
  size,
  ...props
}) => {
  return (
    <Component
      aria-hidden="true"
      className={cn(svgIconVariants({ size, className }))}
      focusable="false"
      {...props}
    />
  );
};

SvgIcon.displayName = 'SvgIcon';
