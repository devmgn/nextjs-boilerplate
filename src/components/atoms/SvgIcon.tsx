import styled from '@emotion/styled';
import createShouldForwardProp from '@/helpers/createShouldForwardProp';
import type { CSSProperties } from 'react';

type SvgIconProps = {
  fontSize?: CSSProperties['fontSize'];
  rotate?: CSSProperties['rotate'];
  color?: CSSProperties['color'];
};

const SvgIcon = styled('svg', {
  shouldForwardProp: createShouldForwardProp('fontSize', 'rotate', 'color'),
})<SvgIconProps>`
  display: inline-block;
  flex-shrink: 0;
  width: 1em;
  height: 1em;
  font-size: ${({ fontSize }) => fontSize || '24px'};
  color: ${({ color }) => color || 'inherit'};
  user-select: none;
  fill: currentColor;
  rotate: ${({ rotate }) => rotate || undefined};
`;

SvgIcon.defaultProps = {
  focusable: false,
  'aria-hidden': true,
};

export default SvgIcon;
