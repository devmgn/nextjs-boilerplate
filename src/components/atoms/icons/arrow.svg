<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M4 13L4 11L16 11L10.5 5.5L11.92 4.08L19.84 12L11.92 19.92L10.5 18.5L16 13L4 13Z" fill="black"/>
</svg>
