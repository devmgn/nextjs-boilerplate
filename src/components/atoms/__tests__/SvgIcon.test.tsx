import { render } from '@testing-library/react';
import { Help } from '@/assets/icons/';
import SvgIcon from '../SvgIcon';

describe('Icon', () => {
  test('スナップショットテスト', () => {
    const { asFragment } = render(<SvgIcon as={Help} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
