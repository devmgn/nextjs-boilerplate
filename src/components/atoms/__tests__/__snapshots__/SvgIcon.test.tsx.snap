// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon スナップショットテスト 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 1em;
  height: 1em;
  font-size: 24px;
  color: inherit;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  fill: currentColor;
}

<svg
    aria-hidden="true"
    class="emotion-0 emotion-1"
    focusable="false"
  />
</DocumentFragment>
`;
